src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,610 @@
+    pub fn position(&self, id: FastThreadId) -> Option<usize> {
+        let start_pos = FastThreadId::hash_idx(&id);
+        let mut idxs = (start_pos..TLS_FALLBACK_MAX_THREADS).chain(0..start_pos);
+        let id_as_usize = usize::from(id);
+        idxs.find(|idx| self.0[*idx].load(core::sync::atomic::Ordering::Relaxed) == id_as_usize)
+    }
+
+    /// Find a free storage index and assign it to `id`.
+    pub fn insert(&self, id: FastThreadId) -> usize {
+        let start_pos = FastThreadId::hash_idx(&id);
+        let mut idxs = (start_pos..TLS_FALLBACK_MAX_THREADS).chain(0..start_pos);
+        let none_as_usize = usize::from(FastThreadId::None);
+        let id_as_usize = usize::from(id);
+        idxs.find(|idx| {
+            let res = self.0[*idx].compare_exchange(
+                none_as_usize,
+                id_as_usize,
+                // Acquire the storage for this index, released in
+                // `global_thread_ids::remove`.
+                core::sync::atomic::Ordering::Acquire,
+                core::sync::atomic::Ordering::Relaxed,
+            );
+            res.is_ok()
+        })
+        .unwrap()
+    }
+}
+
+pub mod global_thread_ids {
+    use super::*;
+
+    static THREAD_IDS: LazyLock<FastThreadIds> =
+        LazyLock::const_new(|| FastThreadIds(core::array::from_fn(|_| AtomicUsize::new(0))));
+
+    pub(crate) fn get(id: FastThreadId) -> Option<usize> {
+        THREAD_IDS.force().position(id)
+    }
+
+    pub(crate) fn get_or_insert(tid: FastThreadId) -> usize {
+        let ids = THREAD_IDS.force();
+        let idx = ids.position(tid);
+        idx.unwrap_or_else(|| ids.insert(tid))
+    }
+
+    pub fn remove(idx: usize) {
+        // Release storage corresponding to this `idx`, acquired in `FastThreadIds::insert`.
+        THREAD_IDS.force().0[idx].store(0, core::sync::atomic::Ordering::Release);
+    }
+}
+
+// The raw byte thread local storage for each thread. Indexes are obtained via
+// `FastThreadIds`.
+#[repr(transparent)]
+struct StoragesType([UnsafeCell<ShimThreadLocalStorage>; TLS_FALLBACK_MAX_THREADS]);
+// SAFETY: normally the UnsafeCell forces this to be !Sync, but we ensure
+// only one thread ever has access to any given UnsafeCell.
+unsafe impl Sync for StoragesType {}
+
+#[cfg(not(miri))]
+mod global_storages {
+    use super::*;
+
+    //  Allocate lazily via `mmap`, to avoid unnecessarily consuming
+    // the memory in processes where we always use native thread local storage.
+    static STORAGES: LazyLock<&'static StoragesType> = LazyLock::const_new(|| {
+        let ptr = unsafe {
+            log::debug!("Allocating fallback thread-local storage");
+            rustix::mm::mmap_anonymous(
+                core::ptr::null_mut(),
+                core::mem::size_of::<StoragesType>(),
+                ProtFlags::READ | ProtFlags::WRITE,
+                MapFlags::PRIVATE,
+            )
+            .unwrap()
+        };
+        unsafe { &*(ptr as *const StoragesType) }
```

Comment:
What is the alignment of the pointer returned from `mmap_anonymous`, and does it have a valid alignment for `StoragesType`? (If so, add some documentation?)